tracer/test/Datadog.Trace.Security.Unit.Tests/FuzzEncoder.cs

```diff
@@ -0,0 +1,193 @@
+// <copyright file="FuzzEncoder.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using Bogus;
+using Datadog.Trace.AppSec.Waf;
+using Datadog.Trace.AppSec.Waf.Initialization;
+using Datadog.Trace.AppSec.Waf.NativeBindings;
+using Datadog.Trace.Vendors.Newtonsoft.Json;
+using Datadog.Trace.Vendors.Newtonsoft.Json.Linq;
+using Xunit;
+
+namespace Datadog.Trace.Security.Unit.Tests
+{
+    public class FuzzEncoder
+    {
+        private static readonly JTokenType[] OpeningTypes = new[] { JTokenType.Array, JTokenType.Object };
+        private static readonly JTokenType[] TerminatingTypes = new[] { JTokenType.Integer, JTokenType.Float, JTokenType.String, JTokenType.Boolean, JTokenType.Null, JTokenType.Date, JTokenType.Guid, JTokenType.Uri, JTokenType.TimeSpan };
+        private readonly Random _rnd = new();
+        private readonly Faker faker = new("en");
+
+        [Fact]
+        [Trait("Category", "ArmUnsupported")]
+        public void LetsFuzz()
+        {
+            // if we don't throw any exceptions and generate a valid object the the test is successful
+
+            var libraryHandle = LibraryLoader.LoadAndGetHandle();
+            var wafNative = new WafNative(libraryHandle);
+            var encoder = new AppSec.Waf.Encoder(wafNative);
+
+            for (int i = 0; i < 100; i++)
+            {
+                var jsonReader = GenerateJson();
+                var root = JToken.ReadFrom(jsonReader);
+
+                var l = new List<Obj>();
+                using var result = encoder.Encode(root, l);
+
+                // check the object is valid
+                Assert.NotEqual(ObjType.Invalid, result.ArgsType);
+
+                l.ForEach(x => x.Dispose());
+            }
+        }
+
+        // uncomment for a quick way to view the gnerated json
+        // [Fact]
+        private void ViewJson()
+        {
+            var memortyStream = GenerateJsonMemoryStream();
+            var jsonText = Encoding.UTF8.GetString(memortyStream.GetBuffer(), 0, (int)memortyStream.Length);
+            Console.WriteLine(jsonText);
+        }
```

Comment:
Not sure this is necessary or works correctly (can't necessarily use Console.WriteLine with XUnit). Instead, you could add it to the `GenerateJsonMemoryStream` and write it to an `ITestOutputHelper` - the output is only shown in CI on test failures, so best of both worlds I think?

Reply:
Yeah, not sure what I was thinking, it would be frustrating to have a test failure and not have access to the the json that caused it. See dea63efa22f52e2aec7b97a34e6186345ac481eb for the fix.